autonat/autonat-v2.md

```diff
@@ -0,0 +1,287 @@
+Requests for verifying such addresses can be reused to get information about
+other addresses
+
+The client can verify that the server did successfully dial an address of the
+same transport as it reported in the response by checking the local address of
+the connection on which the nonce was received on.  
+
+Compared to `autonat v1` there are three major differences
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address.
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+3. `autonat v2` provides a mechanism for nodes to dial an IP address different
+from the requesting node's observed IP address without risking amplification
+attacks. `autonat v1` disallowed such dials to prevent amplification attacks.
+
+
+## AutoNAT V2 Protocol
+
+![Autonat V2 Interaction](autonat-v2.svg)
+
+A client node wishing to determine reachability of its adddresses sends a
+`DialRequest` message to a server on a stream with protocol ID
+`/libp2p/autonat/2/dial`. Each `DialRequest` is sent on a new stream.
+
+This `DialRequest` message has a list of addresses and a fixed64 `nonce`. The
+list is ordered in descending order of priority for verification. AutoNAT V2 is
+only for testing reachability on Public Internet. Client MUST NOT send any
+private address as defined in [RFC
+1918](https://datatracker.ietf.org/doc/html/rfc1918#section-3) in the list. 
+
+Upon receiving this request, the server selects the first address from the list
+of addresses that it is capable of dialing. The server MUST NOT dial any address
+other than this one. If this selected address has an IP address different from
+the requesting node's observed IP address, server initiates the Amplification
+attack prevention mechanism (see [Amplification Attack
+Prevention](#amplification-attack-prevention) ). On completion, the server
+proceeds to the next step. If the selected address has the same IP address as
+the client's observed IP address, server proceeds to the next step skipping
+Amplification Attack Prevention steps.
+
+The server dials the selected address, opens a stream with Protocol ID
+`/libp2p/autonat/2/attempt` and sends a `DialAttempt` message with the nonce
+received in the request. The server MUST close this stream after sending the
+`DialAttempt` message.
+
+Upon completion of the dial attempt, the server sends a `DialResponse` message
+to the client node on the `/libp2p/autonat/2/dial` stream. The response contains
+a list of `DialStatus`es with a status for each address in the list up to and
+including the address that the server attempted to dial. The `DialStatus` for an
+address is set according to [Requirements for
+DialStatus](#requirements-for-dialstatus). The response also contains an
+appropriate `ResponseStatus` set according to [Requirements For
+ResponseStatus](#requirements-for-responsestatus).
```

Comment:
I think a single status is fine too. 
I wanted to specify the reason for refusing an address because that'd have allowed the client to infer more accurately that the particular address that the server refused to dial is a new transport not widely supported in the network. However multiple refusals to probe that address would support the same conclusion on a client.

Reply:
I've changed it to sending a single status. cc @thomaseizinger

Reply:
Works for me. The client probably doesn't need to know why a certain address was skipped. With a high probability, all previous addresses are not supported / understood.

Note that almost the same profiling can be done using the server's identify response and the advertised listen addresses.